[Paul Cheng] iP

itsapaulblem opened the pull request:
AldenPro

>  "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

AldenPro eases your brain of having to remember things you need to do. It's, 

- text-based
- easy to learn 
- ~~FAST~~ SUPER FAST to use 

All you need to do is, 

1.  download it from here [url](https://github.com/itsapaulblem/ip)
2.  double-click it. 
3.  add your tasks
4.  let it manage your tasks for you 👍 

And it is **FREE** !!

Features: 

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [x] Reminders (coming soon)

---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` mthod:

```java
public class Main {
  public static void main(String [ ] args) {
    Application.launch(MainApp.class, args);
  }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

itsapaulblem committed:
Rename chatbot to Alden and implement Level-0: Greet and Exit

itsapaulblem committed:
Implement Level-1: Echo Functionality

itsapaulblem committed:
Implement Level-2: Add and List Functionality

itsapaulblem committed:
Implement Lvel-3: Mark as Done and Unmark Functionality

itsapaulblem committed:
Implement Level-4: Mark ToDo, Event, Deadline Functionality

itsapaulblem committed:
Implement Level-4: Mark ToDo, Event, Deadline Functionality

itsapaulblem committed:
Implement Level-5 Handle Errors

itsapaulblem committed:
Implement Level-6 Delete

itsapaulblem committed:
Added runtest.bat files

itsapaulblem committed:
Implemented Level-7 Save

itsapaulblem committed:
Merge branch 'branch-Level-7'
Following task7 instructions

itsapaulblem committed:
Implemented Level-8 Dates and Times

itsapaulblem committed:
Merge branch 'branch-Level-8'
Following task 8 instructions

itsapaulblem committed:
Made the code more OOP

itsapaulblem committed:
A-Package

itsapaulblem committed:
Merge remote-tracking branch 'origin/add-gradle-support'
Following A-Gradle instructions

itsapaulblem committed:
Merge branch 'A-JUnit' of https://github.com/itsapaulblem/ip into A-JUnit

itsapaulblem committed:
Add JUnit Tests

itsapaulblem committed:
added seedu.duke.main

itsapaulblem committed:
added seedu.duke.Main

itsapaulblem committed:
Updated build.gradle

itsapaulblem committed:
Implemented Level-9 Find

itsapaulblem committed:
Added JavaDocs

itsapaulblem committed:
Fixed Find Bug

itsapaulblem committed:
improved coding standards

itsapaulblem committed:
Merge tag 'A-CodingStandard'

itsapaulblem committed:
Merge tag 'Level-9'

itsapaulblem committed:
Fixed Level-9 Bug

itsapaulblem committed:
Improved the coding standards of all classes

itsapaulblem committed:
Added checkstyle and improved coding standards

itsapaulblem committed:
Improved coding standards using CheckStyle

itsapaulblem committed:
Implemented JavaFx into chatbot

itsapaulblem committed:
Implemented basic UI for Alden Bot

itsapaulblem committed:
improved ui output

itsapaulblem committed:
added photos for user input in dialogue box

itsapaulblem committed:
Added javadocs for new dialogue classes

itsapaulblem committed:
Fixed error exception date format bug on UI

itsapaulblem committed:
Improved coding standard

itsapaulblem committed:
Fixed checkstyle warnings

itsapaulblem committed:
edited package name Alden to alden and fixed checkstyle errors

itsapaulblem committed:
Resolved merge conflicts between master and branch-Level-9

itsapaulblem committed:
Fix accidental mistake of deleting a few java files from previous commit

itsapaulblem committed:
fix checkstyle warning

CJ-783 commented on src/main/java/alden/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,45 @@
+     * @throws AldenException If the user input is malformed or doesn't contain a valid deadline.
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws AldenException {
+        String[] parts = this.userInput.split("/by", 2); // Split the input by the "/by" keyword
```
Perhaps the array name could be more informative?

CJ-783 commented on src/main/java/alden/AddEventCommand.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws AldenException {
+        String[] parts = userInput.split("/from|/to", 3);
```
Perhaps the array name could be more informative?

CJ-783 commented on src/main/java/alden/Main.java:
```diff
@@ -0,0 +1,137 @@
+     * Handles user input from the text field.
+     */
+    private void handleInput() {
+        String input = inputField.getText().trim();
```
Perhaps the variable name "input" could be more informative?

CJ-783 commented on src/main/java/alden/Task.java:
```diff
@@ -0,0 +1,94 @@
+ * This is the base class for different types of tasks (e.g., Todo, Deadline, Event).
+ */
+public abstract class Task {
+    protected String description; // Description of the task
```
Perhaps the variable name "description" could be more informative?

CJ-783 commented on src/main/java/alden/Task.java:
```diff
@@ -0,0 +1,94 @@
+ */
+public abstract class Task {
+    protected String description; // Description of the task
+    protected boolean isDone; // Status of whether the task is completed
```
I like how the variable of "isDone" is made to sound like booleans

CJ-783 reviewed (commented):
LGTM! 🔥 🔥 

Just a few nits to fix.

CJ-783 commented on src/main/java/alden/Alden.java:
```diff
@@ -0,0 +1,53 @@
+        ui.showWelcome();
+
+        // Main input loop
+        while (isRunning) {
+            try {
+                String userInput = scanner.nextLine().trim();
+
+                if (userInput.equalsIgnoreCase("bye")) {
+                    isRunning = false;
+                    ui.showGoodbye();
+                } else {
+                    // Parse the user input into a command and execute it
+                    Command command = Parser.parse(userInput);
+                    command.execute(tasks, ui, storage);
+                }
+            } catch (AldenException e) {
+                // Display error message if there is an exception
+                ui.showError(e.getMessage());
+            }
+        }
+        scanner.close();
+    }
```
I like the use of K&R style bracket in the while loop!

CJ-783 commented on src/main/java/alden/Parser.java:
```diff
@@ -0,0 +1,44 @@
+     * @throws AldenException If the command is invalid or unrecognized.
+     */
+    public static Command parse(String fullCommand) throws AldenException {
+        if (fullCommand.startsWith("todo")) {
+            return new AddTodoCommand(fullCommand); // Create a new AddTodoCommand
+        } else if (fullCommand.startsWith("deadline")) {
+            return new AddDeadlineCommand(fullCommand); // Create a new AddDeadlineCommand
+        } else if (fullCommand.startsWith("event")) {
+            return new AddEventCommand(fullCommand); // Create a new AddEventCommand
+        } else if (fullCommand.startsWith("mark")) {
+            return new MarkTaskCommand(fullCommand, true); // Create a new MarkTaskCommand to mark as done
+        } else if (fullCommand.startsWith("unmarked")) {
+            return new MarkTaskCommand(fullCommand, false); // Create a new MarkTaskCommand to unmark as done
+        } else if (fullCommand.startsWith("delete")) {
+            return new DeleteTaskCommand(fullCommand); // Create a new DeleteTaskCommand
+        } else if (fullCommand.equalsIgnoreCase("list")) {
+            return new ListTasksCommand(); // Create a new ListTasksCommand
+        } else if (fullCommand.equalsIgnoreCase("bye")) {
+            return new ExitCommand();
+        } else if (fullCommand.startsWith("find")) {
+            return new FindTaskCommand(fullCommand);
+        } else if (fullCommand.isEmpty()) {
+            throw new AldenException("Keyword cannot be empty for find command.");
+        } else {
+            throw new AldenException("Invalid command: " + fullCommand); // Throw exception if the command is invalid
+        }
```
I like how neat the if-else statement (form) is written.

CJ-783 commented on src/main/java/alden/Storage.java:
```diff
@@ -0,0 +1,107 @@
+
+                String[] parts = line.split(" \\| ");
+                String taskType = parts[0];
+                boolean isDone = "1".equals(parts[1]);
```
I like how the boolean is named like booleans

CJ-783 commented on src/main/java/alden/Task.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Perhaps the name of this boolean method could sound more like booleans?

CJ-783 reviewed (commented):
LGTM! 🔥 🔥 🔥 

Just some nits to fix

itsapaulblem committed:
Add assertions to verify code assumptions

Add assertions to verify and document code assumptions

The code previously lacked explicit documentation of assumptions and invariants.
This change adds runtime assertions throughout the codebase to:

* Verify constructor preconditions for non-null and valid inputs
* Maintain task list integrity after modifications
* Enforce temporal logic in events and deadlines
* Validate data format consistency in storage operations
* Check method preconditions for critical operations

These assertions serve dual purposes:
1. Act as executable documentation of code assumptions
2. Catch programming errors early during development

The assertions are added to the following classes:
- AddDeadlineCommand: Verify task creation and list updates
- AddEventCommand: Validate event time constraints
- TaskList: Ensure list operation integrity
- Storage: Check data format consistency
- Deadline/Event: Enforce temporal constraints

Note that assertions must be enabled with -ea flag during development.
They should not be relied upon in production as they may be disabled.

Add invariant checks for TaskList operations

Added assertions to TaskList class operations to maintain data integrity:

* Validate index bounds before access/modification
* Verify task presence after addition
* Confirm task removal after deletion
* Check search keyword validity

These checks help catch potential issues during development and
serve as documentation of the TaskList class's invariants.

Add temporal validation assertions to Event and Deadline

Added runtime checks to enforce temporal constraints:

* Event end time must be after start time
* Deadlines cannot be set in the past
* DateTime objects must be properly initialized
* Date/time string formats must be valid

These assertions help prevent invalid temporal states and
document the expected behavior of time-based tasks.

itsapaulblem committed:
Refactor and Improve code quality

Add AddDeadlineCommand to handle deadline task input

The code previously lacked proper handling of deadline task input and did not adhere to the CS2103T code quality specifications. This commit introduces the AddDeadlineCommand class, which processes user input to add a deadline task to the task list. The class expects the input to follow a specific format, containing a description and a "/by" clause to specify the deadline time. If the input is malformed (e.g., missing "/by" or description), an exception is thrown. Additionally, the task list is updated and saved to storage after adding the new task.

This change:

* Implements logic to extract description and deadline from user input
* Updates the task list and saves the changes to storage

These improvements ensure that the deadline tasks are handled consistently and that user input is validated correctly, improving the maintainability of the codebase.

Add date parsing fix in Deadline class for "today" and "tomorrow" keywords

The code previously had issues with parsing special keywords like "today" and "tomorrow" in the Deadline class, causing errors when processing user input. This commit addresses the issue by ensuring that these keywords are correctly interpreted, setting the deadline to the current date or the next day.

The fix includes:

* Corrected date parsing logic in the Deadline class to handle "today" and "tomorrow"
* Updated the DateTimeFormatter to support both date and date-time formats

This change allows users to specify deadlines using relative dates like "today" and "tomorrow" without errors, making the task management more intuitive.

Improve documentation for AddEventCommand and AddTodoCommand classes

This commit improves the documentation for the AddEventCommand and AddTodoCommand classes. The documentation now clarifies the expected input formats, input validation, and exception handling for both command classes. This change ensures that other developers can easily understand the purpose of each class and how to use them effectively.

Changes made:

* Added explanations for input formats and exception handling in both classes
* Updated method-level comments to provide more detailed information
* Improved code clarity and readability

This update enhances the code maintainability and provides better guidance for future developers.

itsapaulblem committed:
Merge pull request #1 from itsapaulblem/A-Assertions

Add assertions to verify code assumptions

itsapaulblem committed:
Merge branch 'master' into A-CodeQuality

itsapaulblem committed:
Merge pull request #2 from itsapaulblem/A-CodeQuality

Refactor and Improve code quality

itsapaulblem committed:
Fix AddEventCommand to correctly handle task creation and adding to task list

itsapaulblem committed:
refactor: remove unnecessary comment

itsapaulblem committed:
Refactor: Align code with checkstyleTest standards

itsapaulblem committed:
Refactor: Align code with checkstyleMain standards

itsapaulblem committed:
Refactor: Align code with checkstyleMain standards

itsapaulblem committed:
Refactor: Remove unnecessary comment

itsapaulblem committed:
Refactor: Remove unnecessary comment

itsapaulblem committed:
Fix: Correct test cases with future date expectations

itsapaulblem committed:
feat: Add chronological sorting for tasks

- Implement SortCommand to enable sorting tasks by date
- Support chronological sorting for Deadline and Event tasks
- Modify Parser to recognize new 'sort' command
- Create sorting mechanism that preserves Todo task order
- Add getStartDateTime() and getEndDateTime() methods to Event class
- Provide user-friendly sorting with date-based comparator
- Save sorted task list to storage after sorting

itsapaulblem committed:
Revert "Feat: Add chronological sorting for tasks"

This reverts commit d26eed9bb1e818c86e515e281893e27dccf00e25.

itsapaulblem committed:
Feat: Add chronological sorting for tasks

- Implement SortCommand to enable sorting tasks by date
- Support chronological sorting for Deadline and Event tasks
- Modify Parser to recognize new 'sort' command
- Create sorting mechanism that preserves Todo task order
- Add getStartDateTime() and getEndDateTime() methods to Event class
- Provide user-friendly sorting with date-based comparator
- Save sorted task list to storage after sorting

itsapaulblem committed:
Refactor: Align code with checkstyleMain standards

itsapaulblem committed:
Refactor: Align code with checkstyleMain standards

itsapaulblem committed:
Merge pull request #3 from itsapaulblem/BCD-Extension

Feat: Implement Task Sorting Functionality

itsapaulblem committed:
Docs: Add Ui.png to Documentation

itsapaulblem committed:
Update README.md

itsapaulblem committed:
Refactor(parser): Edit Unmark command handling

itsapaulblem committed:
Merge branch 'master' of https://github.com/itsapaulblem/ip

itsapaulblem committed:
Fix: Configure shadowJar and JavaFX dependencies

itsapaulblem committed:
Update README.md

xplus2g4 commented on config/checkstyle/suppressions.xml:
```diff
@@ -0,0 +1,10 @@
+<suppressions>
+    <suppress checks="JavadocType" files=".*Test\.java"/>
+    <suppress checks="MissingJavadocMethodCheck" files=".*Test\.java"/>
+</suppressions>
```
It might be good if you leave a new line at the end of the file.
See: https://stackoverflow.com/a/2287990

xplus2g4 commented on test-jar/alden.jar:
Is this committed by accident?

xplus2g4 commented on test-jar/data/Alden.txt:
Is this committed by accident?

xplus2g4 commented on src/main/java/alden/TaskList.java:
```diff
@@ -0,0 +1,106 @@
+     * @return An ArrayList of tasks that match the given keyword.
+     */
+    public ArrayList<Task> findTasks(String keyword) {
+        assert keyword != null && !keyword.trim().isEmpty() : "Search keyword cannot be null or empty";
```
Maybe you can handle the error instead of asserting it.

xplus2g4 commented on src/main/java/alden/TaskList.java:
```diff
@@ -0,0 +1,106 @@
+     * @return The size of the task list.
+     */
+    public int size() {
+        assert tasks.size() >= 0 : "Task list size cannot be negative";
```
Probably redundant, `ArrayList` should return a non-negative integer. (I didn't verify it, but I believe so.)

xplus2g4 commented on src/main/java/alden/TaskList.java:
```diff
@@ -0,0 +1,106 @@
+        assert index >= 0 && index < tasks.size() : "Index must be within bounds";
+        Task taskToRemove = tasks.get(index);
+        Task removed = tasks.remove(index);
+        assert removed.equals(taskToRemove) : "Removed task should match requested task";
```
Handled by `ArrayList`, you probably don't need to assert this.

xplus2g4 commented on src/main/java/alden/TaskList.java:
```diff
@@ -0,0 +1,106 @@
+        assert task != null : "Cannot add null task";
+        int sizeBefore = tasks.size();
+        tasks.add(task);
+        assert tasks.size() == sizeBefore + 1 : "Task list size should increase by 1";
```
Handled by ArrayList, you probably don't need to assert this.

xplus2g4 commented on src/main/java/alden/TaskList.java:
```diff
@@ -0,0 +1,106 @@
+        int sizeBefore = tasks.size();
+        tasks.add(task);
+        assert tasks.size() == sizeBefore + 1 : "Task list size should increase by 1";
+        assert tasks.get(tasks.size() - 1).equals(task) : "Added task must be in list";
```
In other ListLike data structures, a new element may not appear at the end.

xplus2g4 commented on src/main/java/alden/Storage.java:
```diff
@@ -0,0 +1,110 @@
+        try {
+            File file = new File(filePath);
+            if (!file.exists()) {
+                file.getParentFile().mkdirs(); // Create directories if they do not exist
```
Nice catch of edge cases

xplus2g4 reviewed (commented):
LGTM, requested some cosmetic changes. On a side note, maybe you can try to organize your code in folders for better navigation. Well done!

xplus2g4 commented on src/main/java/alden/Main.java:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * Handles user input from the text field.
+     */
+    private void handleInput() {
```
Can refactor and flatten this code

itsapaulblem committed:
Refactor: Improve code quality

itsapaulblem committed:
Refactor: Improve the coding standard of headers

itsapaulblem committed:
Refactor: Apply SLAP to Storage and Main classes

itsapaulblem committed:
Refactor: Improve code quality

itsapaulblem committed:
Update README.md

itsapaulblem committed:
Update README.md

itsapaulblem committed:
Debug: Remove unnamed module warning

itsapaulblem committed:
Refactor: Build gradle file
